RUFAS/biophysical/manure/storage/compost_bedded_pack_barn.py

```
@@ -0,0 +1,307 @@
+        self._manure_to_process.degradable_volatile_solids = degradable_volatile_solids_after_losses
+        self._manure_to_process.total_solids = total_solids_after_losses
+
+    def _apply_nitrogen_losses(
+        self, storage_nitrous_oxide_N: float, storage_ammonia_N: float, storage_N_loss_from_leaching: float
+    ) -> None:
+        """
+        This function applies the nitrogen losses to the received manure nitrogen and ammoniacal nitrogen in place.
+
+        Parameters
+        ----------
+        storage_nitrous_oxide_N : float
+            The nitrogen loss through nitrous oxide emissions on the current day, kg.
+        storage_ammonia_N : float
+            The nitrogen loss through ammonia emissions on the current day, kg.
+        storage_N_loss_from_leaching : float
+            The nitrogen loss through leaching on the current day, kg.
+
+        Raises
+        ------
+        ValueError
+            If the total nitrogen losses are greater than the total received manure nitrogen.
+        """
+        received_manure_nitrogen_after_losses = (
+            self._manure_to_process.nitrogen
+            - storage_nitrous_oxide_N
+            - storage_ammonia_N
+            - storage_N_loss_from_leaching
+        )
+        if received_manure_nitrogen_after_losses < 0:
+            self._om.add_error(
+                "Nitrogen loss application error",
+                "Cannot have total nitrogen losses greater than total received manure nitrogen.",
+                info_map={"class": self.__class__.__name__, "function": self._apply_nitrogen_losses.__name__},
+            )
+            raise ValueError(
+                "Nitrogen loss application error: cannot have total nitrogen losses greater than "
+                "total received manure nitrogen."
+            )
+        self._manure_to_process.ammoniacal_nitrogen = max(
+            0.0, self._manure_to_process.ammoniacal_nitrogen - storage_ammonia_N
+        )
```

Do we also want to raise an error here if the ammoniacal nitrogen after loss is negative? Or just capping it at 0.0?

---

We should.

---

@elle-andreen said specifically the max() check would be preferable to raising an error for ammonia losses. The max check matches how it's implemented in Open Lot and Composting.

---

Yes, we want to do _only_ the max() check here, not return an error if storage ammonia > manure ammoniacal nitrogen. It is mathematically possible that we could have ammonia emissions greater than the ammoniacal nitrogen value, because ammonia emissions in CBPB are based on total nitrogen, not ammoniacal nitrogen. I know it's a little confusing, even I have to remind myself why this is the case every time this question comes up!